### ISSUE TITLE
Generic `RowDataType` usability (introduced in #418)

### ISSUE THREAD

**Issue by icecrime** [rsuite/rsuite-table#422]
### Versions

| package        | version |
| -------------- | ------- |
| `react`        | `18.2.0` |
| `rsuite-table` | `5.10.3` |

### What is the current behavior?

_I originally posted [this comment](https://github.com/rsuite/rsuite-table/pull/418#issuecomment-1503547666) on the pull request itself but I'm filing an issue for visibility._

The changes introduced by #418 on the `RowDataType` type breaks our build in ways I don't fully understand.

We are using [rsuite's ItemDataType](https://github.com/rsuite/rsuite/blob/main/src/%40types/common.ts#L218) for our tables, but this is not possible anymore:

```
TS2322: Type 'ItemDataType<string | number>[]' is not assignable to type 'readonly RowDataType<never>[]'.
  Type 'ItemDataType<string | number>' is not assignable to type 'RowDataType<never>'.
    Types of property 'children' are incompatible.
      Type 'ItemDataType<string | number>[] | undefined' is not assignable to type 'never[] | undefined'.
```

I wonder whether:

* It's desired for the generic type to default to never?
* It's desired for the children property to be `T[]` rather than a `RowDataType<T>[]`?

Thank you for your work and for your help.

**Comment by Pinpickle** [comment]
I _think_ I'm running into the same issue, just with a different error (the commonality being `never`).

From what I can tell, this is caused here:

https://github.com/rsuite/rsuite-table/blob/dc78cee02f2e05b58599c073b72f1b8e5859a260/src/Table.tsx#L260

The constraint for `Row` is `RowDataType`. `RowDataType`'s default type argument is `never`:

https://github.com/rsuite/rsuite-table/blob/dc78cee02f2e05b58599c073b72f1b8e5859a260/src/%40types/common.ts#L23

So this means, the constraint expands to `Row extends RowDataType<never>`, which expands to

```
Row extends {
  dataKey?: string;
  children?: never[];
  [key: string]: any;
}
```

Nothing extends `never[] | undefined` other than `never[]` or `undefined` (as it's the bottom type). Any type that has a children value set will not work.

Changing this to `unknown` should work:

```diff
- const Table = React.forwardRef(<Row extends RowDataType, Key>(props: TableProps<Row, Key>, ref) => {
+ const Table = React.forwardRef(<Row extends RowDataType<unknown>, Key>(props: TableProps<Row, Key>, ref) => {
```

**Comment by AdrienFromToulouse** [comment]
@simonguo would you agree on this 👆 ?

**Comment by simonguo** [comment]
Hi, all 

Could you provide a minimal example where the above problem occurs?

In addition, the latest version v5.18.1 supports getting the type-inferred Cell from children. Can this solve your problem?
> https://github.com/rsuite/rsuite-table#type-safety

**Comment by simonguo** [comment]
Fixed in v5.18.2
